protocol/network/6-interfaces.md

```diff
@@ -0,0 +1,137 @@
+6 Interfaces
+============
+
+## 6.1 libp2p
+
+## 6.2 Peer Routing
+
+## 6.3 Swarm
+
+~~The network is abstracted through the swarm which presents a simplified interface for the remaining layers to have access to the network. This interface should look like:~~
+
+- `.openStream(peer, protocol)` - peer should contain the ID of the peer and its respective multiaddrs known.
+- `.registerHandler(protocol, handlerFunc)` - enable a protocol to be registered, so that another peer can open a stream to talk with us to that specific protocol
+- `.listen()` - to start listening for incoming connections and therefore opening of streams
```

Comment:
- :+1: on this interface.

how about the below. This assumes that we don't sink protocol muxing into the Swarm, but again, that's up for discussion as it may be relevant to. regardless, we want users to use a libp2p.Node, not the Swarm directly.

```go
// Node is the basic interface of libp2p. It represents a "Host" in the libp2p internet.
// It is capable of opening streams (either as a dialer or a listener) to other Nodes.
//
// The libp2p.Node uses a libp2p.Swarm internally to handle all connections and
// stream multiplexing.
//
// It is meant to have user protocols "mounted" onto the Node (as one would bind
// a socket and Listen, in a regular TCP/IP network stack). This is done with either
// the OpenStream or SetHandler functions (which represent Dialing or Listening) 
// for stream of a specific protocol. Only one "client process" may mount a protocol,
// that means that only one Handler is registered per protocol.ID. This tends to be
// the convenient thing to do, but may be up for discussion (e.g. being part of 
// multiple IPFS-DHTs, etc).
//
// The p2p.Node uses internal protocols (which are mounted on top, like any other
// protocol) such as Identify, NAT (traversal), and Relay, to achieve full connectivity
// on the networks it belongs to.
type p2p.Node interface {
  // OpenStream opens a reliable stream to peer.ID with given protocol.ID. 
  // If there is no prior connection to peer.ID, p2p.Node will attempt to establish one.
  OpenStream(peer.ID, protocol.ID)

  // SetHandler registers a function to be called whenever another peer opens a stream
  // with that specific protocol.ID. If HandlerFunc is nil (default), the handler writes an 
  // ProtocolNotFoundError back and closes the stream.
  SetHandler(protocol.ID, protocol.HandlerFunc)
  
  // Listen starts listening for incoming connections and therefore opening of streams
  // (this may need to handle the addresses, i'll look again at how go-libp2p does it).
  Listen()

  // this will need more things here, will look into go-libp2p to figure this out.
}
```